src/main/java/picard/illumina/parser/readers/CbclReader.java

```diff
@@ -57,117 +63,178 @@
-                    int tileNum = headerBuffer.getInt();
-                    int numClustersInTile = headerBuffer.getInt();
-                    int uncompressedBlockSize = headerBuffer.getInt();
-                    int compressedBlockSize = headerBuffer.getInt();
-                    tileInfo[j] = new TileData(tileNum, numClustersInTile, uncompressedBlockSize, compressedBlockSize);
+                    cycleData[i] = new CycleData(version, headerSize, bitsPerBasecall, bitsPerQualityScore, numberOfBins, qualityBins, numTiles, tileInfo, pfExcluded);
+                    this.streams[i] = stream;
+                    this.streamFiles[i] = bclFile;
+                    byteBuffer.clear();
+                    headerBuffer.clear();
                 }
+            }
+            int totalCycleCount = 0;
 
-                boolean pfExcluded = headerBuffer.get() == 1;
-                cycleData[i] = new CycleData(version, headerSize, bitsPerBasecall, bitsPerQualityScore, numberOfBins, qualityBins, numTiles, tileInfo, pfExcluded);
-                this.streams[i] = stream;
-                this.streamFiles[i] = bclFile;
-                byteBuffer.clear();
-                headerBuffer.clear();
+            if (cycleData[totalCycleCount].tileInfo == null) {
+                throw new PicardException("Could not find tile " + cycleData[totalCycleCount].tileInfo);
             }
 
+            for (int outputLength : outputLengths) {
+                for (int cycle = 0; cycle < outputLength; cycle++) {
+                    CycleData currentCycleData = cycleData[totalCycleCount];
+                    try {
+                        if (cachedTile[totalCycleCount] == null) {
+                            if (!cachedFilter.containsKey(cycleData[totalCycleCount].tileInfo.tileNum)) {
+                                cacheFilterAndLocs(cycleData[totalCycleCount].tileInfo, locs);
+                            }
+                            cacheTile(totalCycleCount, cycleData[totalCycleCount].tileInfo, currentCycleData);
+                        }
+                    } catch (IOException e) {
+                        // when logging the error, increment cycle by 1, since totalCycleCount is zero-indexed but Illumina directories are 1-indexed.
+                        throw new PicardException(String.format("Error while reading from BCL file for cycle %d. Offending file on disk is %s",
+                                (totalCycleCount + 1), this.streamFiles[totalCycleCount].getAbsolutePath()), e);
+                    }
+                    totalCycleCount++;
+                }
+            }
 
         } catch (final IOException ioe) {
             throw new RuntimeIOException(ioe);
         }
     }
 
+    private Map<Integer, Map<Integer, File>> sortCbcls(List<File> cbcls) {
+        Map<Integer, Map<Integer, File>> sortedMap = new TreeMap<>();
+        Pattern pattern = Pattern.compile("^.+C(\\d{1,4}).+L(\\d{1,3})_(\\d).cbcl$");
```

Comment:
The result of `Pattern.compile()` is thread safe so it can be stored in a static final field to avoid doing it more than once.